✨ New WordPress.WP.OptionAutoload sniff

rodrigoprimo opened the pull request:
This sniff warns users about missing `$autoload` parameter when calling relevant WordPress option functions.

It includes tests and sniff documentation in the XML format.

Closes #2473 

### Sniff behavior

The sniff runs when the following functions are found in the code:
  - add_option( $option, $value = '', $deprecated = '', $autoload = null)
  - update_option( $option, $value, $autoload = null )
  - wp_set_options_autoload( $options, $autoload )
  - wp_set_option_autoload( $option, $autoload )
  - wp_set_option_autoload_values( $options )

For `add_option()` and `update_option()`, if the `$autoload` parameter is not passed, it simply recommends passing it. For the other three functions, the `$autoload` parameter (or `$options` parameter in the case of `wp_set_option_autoload_values()`) is mandatory, so it is not needed to check if the parameter is omitted.

If the `$autoload` parameter is passed and is `true|false|null` for `add_option()` and `update_option()` or `true|false` for `wp_set_options_autoload()`, `wp_set_option_autoload()` and `wp_set_option_autoload_values()`, the sniff stays silent as those are valid values.

If the `$autoload` parameter is passed and is `'yes'|'no'`, the sniff flags as discouraged parameter values and auto-fixes to `true|false`.

For the internal-use only values of the `$autoload` parameter, if the value is `'auto'|'auto-on'|'auto-off'` the sniff flags it as unsupported without auto-fixing. If the value is `'on'|'off'`, the sniff flags it as unsupported and auto-fixes to `true|false`.

Any other value passed is flagged as unsupported, and a list of supported values is provided.

### Open questions

I'm unsure about a few choices that I made in the XML documentation and could use some feedback on the following:

- I opted to include a link to a document explaining how to decide which value to use for the `$autoload` parameter. Since I could not find something wp.org, I ended up linking to a post written by a core committer in their blog. I'm not sure if this is a problem or not.
- Maybe combining the `<standard>` blocks for invalid, deprecated, and internal-only values in a single block would be better? I created a block for each, but there is some repetition in the block description and the valid examples.
- I added a `<standard>` block without code examples at the top with a general description of what the sniff does before creating blocks for each of the warnings generated by the sniff.

rodrigoprimo committed:
✨ New WordPress.WP.OptionAutoload sniff

This sniff warns users about missing `$autoload` parameter when calling relevant WordPress option functions.

The sniff runs when the following functions are found in the code:
  - add_option( $option, $value = '', $deprecated = '', $autoload = null)
  - update_option( $option, $value, $autoload = null )
  - wp_set_options_autoload( $options, $autoload )
  - wp_set_option_autoload( $option, $autoload )
  - wp_set_option_autoload_values( $options )

For `add_option()` and `update_option()`, if the `$autoload` parameter is not passed, it simply recommends passing it. For the other three functions, the `$autoload` parameter (or `$options` parameter in the case of `wp_set_option_autoload_values()`) is mandatory, so it is not needed to check if the parameter is omitted.

If the `$autoload` parameter is passed and is `true|false|null` for `add_option()` and `update_option()` or `true|false` for `wp_set_options_autoload()`, `wp_set_option_autoload()` and `wp_set_option_autoload_values()`, the sniff stays silent as those are valid values.

If the `$autoload` parameter is passed and is `'yes'|'no'`, the sniff flags as discouraged parameter values and auto-fixes to `true|false`.

For the internal-use only values of the `$autoload` parameter, if the value is `'auto'|'auto-on'|'auto-off'` the sniff flags it as unsupported without auto-fixing. If the value is `'on'|'off'`, the sniff flags it as unsupported and auto-fixes to `true|false`.

Any other value passed is flagged as an unsupported value and a list of supported values is provided.

The sniff ignores the function call when it is not able to determine the value of the `$autoload` parameter. For example, when a variable or a constant is passed.

jrfnl reviewed (changes requested):
@rodrigoprimo Thank you for working on this.

I've so far only done a first pass reading through it (without running the sniff/docs etc), but seeing the amount of remarks I've already left, I figured I should give you the chance to iterate on the current state before continuing my review.

rodrigoprimo committed:
Apply suggestions from code review

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,135 @@
+    <![CDATA[
+    When using `add_option()`, `update_option()`, `wp_set_options_autoload()`,
+    `wp_set_option_autoload()`, or `wp_set_option_autoload_values()`, it is recommended to
+    explicitly set the autoload value to ensure predictable behavior. This value determines whether
```
I tried not to use `parameter` as for `wp_set_option_autoload_values()` autoload is not a parameter. But I agree that value is probably worst. I will accept your suggestion, and I will let you know if I can think of a better approach.

rodrigoprimo committed:
Attemp to improve the XML documentation by combining standards

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,416 @@
+
+	/**
+	 * The phrase to use for the metric recorded by this sniff.
+	 *
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,416 @@
+	 *
+	 * @var string[]
+	 */
+	protected static $valid_values_add_and_update = array( 'true', 'false', 'null' );
```
I changed all the `protected` properties and methods to `private`, with the exception of `$group_name` and `$target_functions`, as their access level should be the same or weaker than the corresponding properties in the parent class.

I also changed the `static` properties to regular properties.

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,416 @@
+		$this->phpcsFile->recordMetric( $stackPtr, self::METRIC_NAME, 'param missing' );
+
+		// Only display a warning for the functions in which the `$autoload` parameter is optional.
+		if ( in_array( $function_name, self::$autoload_is_optional, true ) ) {
```
Good point. I believe we discussed this already in the past and I failed to take it into account while creating this sniff. I will think a bit more about how to implement this in the other places and get back to you.

rodrigoprimo replied:
> Please change the $autoload_is_optional array to 'name' => true format and check using isset() for a small performance improvement.
> ☝🏻 This second point should be fixed in quite a few places in this sniff.

I implemented this change in all the places that seemed relevant to me. For `$valid_values_add_and_update` and `$valid_values_other_functions`, I used the same format as what is used in the `\PHPCSUtils\Tokens\Collections` properties and repeated the value in the index of each element of the array. I opted to do this to avoid calling `array_keys()` when passing `$valid_values` to `array_map()`:

https://github.com/WordPress/WordPress-Coding-Standards/blob/fc98b8eb338a18e71b8bab9fd26674bbbaa9733f/WordPress/Sniffs/WP/OptionAutoloadSniff.php#L404-L409

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,416 @@
+	 * @return void
+	 */
+	protected function maybe_display_missing_autoload_warning( $stackPtr, $function_name ) {
+		$this->phpcsFile->recordMetric( $stackPtr, self::METRIC_NAME, 'param missing' );
```
I'm not sure about it as well.

rodrigoprimo replied:
Done

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,135 @@
+    the option is automatically loaded on every page load, which can impact site performance.
+
+    Check https://felix-arntz.me/blog/autoloading-wordpress-options-efficiently-and-responsibly/ for
+    more information on when to set autoload to `true` or `false`.
```
Removed

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,135 @@
+    <![CDATA[
+    Only `true` and `false` are valid values for the autoload parameter, with one exception: `null`
+    is also valid for `add_option()` and `update_option()`.
+    ]]>
```
In a separate commit, I tried again to structure the XML documentation. I kept the "introductory" standard block and the block for the presence of the autoload parameter. The other three blocks were combined into a single one that describes the autoload parameter's valid and invalid values. Any input on this new structure is welcome. I'm unsure if I should combine everything into a single `<standard>` and if I'm insisting too much on separating the parameter is missing warning from the other warnings.

jrfnl replied:
I do think the "missing parameter" and the "invalid parameter value" issues are two separate issues, but I'm not convinced the changed structure of the docs makes this clear enough.

If I look at the latest version, my recommendation would be:
* Move the last `code_comparison` to be directly under the first `standard` block (and drop the third `standard` block which was originally associated with that `code_comparison`.

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,135 @@
+    It is not recommended to use internal-only values ('auto', 'auto-on', 'auto-off', 'on', 'off')
+    in plugin or theme code. Instead, use `true`|`false`|`null` if using `add_option()` or
+    `update_option()` or `true`|`false` if using `wp_set_option_autoload_values()`,
+    `wp_set_option_autoload()`, or `wp_set_options_autoload()`.
```
See https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r1975996912

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,135 @@
+    <![CDATA[
+    Using 'yes' or 'no' is not recommended. Those values are deprecated. Instead, use
+    `true`|`false`|`null` if using `add_option()` or `update_option()` or `true`|`false` if using
+    `wp_set_option_autoload_values()`, `wp_set_option_autoload()`, or `wp_set_options_autoload()`.
```
See https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r1975996912

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,135 @@
+    Even though the autoload parameter is optional for `add_option()` and `update_option()`, it is
+    recommended to always explicitly set it.
+    ]]>
+    </standard>
```
See https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r1975996912

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,416 @@
+	 * @return void
+	 */
+	protected function check_autoload_value( array $autoload_info, $function_name ) {
+		// Find the first and second param non-empty tokens (the second token might not exist).
```
Removed

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,416 @@
+		}
+
+		if ( $param_second_token
+			&& ! in_array( $this->tokens[ $param_first_token ]['code'], array( T_ARRAY, T_OPEN_SHORT_ARRAY ), true )
```
Done

rodrigoprimo commented on WordPress/Tests/WP/OptionAutoloadUnitTest.inc.fixed:
```diff
@@ -0,0 +1,214 @@
+$ok = add_option( 'my_option', autoload: FALSE, value: 'my_value' );
+$ok = add_option( 'my_option', autoload: false );
+$ok = add_option( autoload: false, deprecated: '', value: 'my_value', option: 'my_option' );
+$ok = \update_option( 'my_option', 'new_value', \true );
+$ok = update_option( 'my_option', 'new_value', \false );
+$ok = update_option( 'my_option', 'new_value', \NULL );
```
While scanning the WP.org plugins codebase, I found code samples passing a "namespaced" boolean as the value of the `$autoload` parameter of some of the option functions. As far as I can test, PHP treats `\true` as `true` (and so on):

https://3v4l.org/KmT8M

So, I thought it was worth adding those test cases and also supporting it in the sniff. What do you think?

dingo-d replied:
Shouldn't these be flagged by the Generic.PHP.LowerCaseConstant sniff?

jrfnl replied:
> So, I thought it was worth adding those test cases and also supporting it in the sniff. What do you think?

@rodrigoprimo Fair point. Agreed.

> Shouldn't these be flagged by the Generic.PHP.LowerCaseConstant sniff?

@dingo-d The `NULL` will be flagged by that sniff, but the `\` prefix will not be flagged as that sniff doesn't look for that.
Might not be a bad idea to create a new sniff for that (PHPCSExtra is probably the best place for that). Happy to see an issue about this in that repo.

Aside from that, these are valid test cases for _this_ sniff, independently of whether another sniff would flag this as sniffs should function independently of each other.

jrfnl replied:
Looking at this now... there looks to be a tokenizer issue with `\true` etc when used on PHP 8.0 or higher.... hmm... I think that needs a fix in PHPCS.
I haven't checked how this sniff handles this....

jrfnl replied:
Oh sod it, wrote a sniff now to forbid this: https://github.com/PHPCSStandards/PHPCSExtra/pull/327
Once merged & released, we may want to consider adding this new sniff to WPCS.

As for the PHPCS tokenizer issue, I've looked into that and decided against fixing it as things will change again in PHPCS 4.0, so fixing it now will only make things more difficult for PHPCS 4.0.

rodrigoprimo commented on WordPress/Tests/WP/OptionAutoloadUnitTest.inc.fixed:
```diff
@@ -0,0 +1,214 @@
+$ok = update_option( 'my_option', autoload: TrUe, value: 'new_value' );
+$ok = update_option( autoload: false, value: 'my_value', option: 'my_option' );
+$ok = update_option( $option_name, get_option( $option_name ), false );
+$ok = wp_set_options_autoload( array( 'option1', 'option2' ), \false );
+$ok = wp_set_options_autoload( array( 'option1', 'option2' ), \true );
```
See https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r1975999191

rodrigoprimo commented on WordPress/Tests/WP/OptionAutoloadUnitTest.inc.fixed:
```diff
@@ -0,0 +1,214 @@
+
+/*
+ * These are ignored because it is not easy to determine the option names and the autoload values or
+ * the `$options` parameter is empty.
```
Ok

rodrigoprimo commented on WordPress/Tests/WP/OptionAutoloadUnitTest.inc.fixed:
I will create more tests and push a new commit in the next few days.

jrfnl replied:
I don't think you need _more_ tests. Probably just updating some of the existing tests to include those variations would be enough.

rodrigoprimo replied:
I updated some of the tests to add more variation in whitespace, comments/annotations in unexpected places, and function names using different cases.

rodrigoprimo commented:
Thanks for your review, @dingo-d and @jrfnl! I believe I addressed all of your comments except two that are still pending, and I will need a bit more time (https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r1974686099 and https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r1974641122).

> The only problem is that that is not official documentation. And if their site goes down, you have a dead link in the documentation.

That makes sense. As @jrfnl suggested, I removed the paragraph that included this link.

> Does the validation passes against the [phpcsdocs.xsd](https://github.com/PHPCSStandards/PHPCSDevTools/blob/develop/DocsXsd/phpcsdocs.xsd)?

I believe so. This is how I tested:

```
$ xmllint --schema phpcsdocs.xsd WordPress/Docs/WP/OptionAutoloadStandard.xml --noout 
WordPress/Docs/WP/OptionAutoloadStandard.xml validates
```

Let me know if I'm missing something.

rodrigoprimo committed:
Fix unintentional coding standard violation in a test case

rodrigoprimo committed:
Update some tests to have more function name case variations

rodrigoprimo committed:
Update some tests to have more whitespaces, comments and annotations

rodrigoprimo committed:
Improve sniff performance by changing the format of some properties

This allows to use isset() instead of in_array()

rodrigoprimo committed:
Use `false === isset()` instead of `! isset()` for consistency with the rest of the sniff code

rodrigoprimo commented:
@jrfnl, I addressed the two points that were pending. Could you please take another look at this PR when you get a chance?

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,102 @@
+    'on', or 'off' in plugin/theme code is strongly discouraged as those values are meant for
+    internal-use only.
+
+    Any other values, including `null` for functions other than `add_option()` and `update_option()`
+    are invalid.
```
```suggestion
    Any other values, including `null` for functions other than `add_option()` and `update_option()`,
    are invalid.
```

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+
+		$normalized_value = strtolower( $autoload_info['clean'] );
+
+		if ( \T_NS_SEPARATOR === $this->tokens[ $param_first_token ]['code'] && $param_second_token ) {
```
Note: this needs to be changed when adding support for PHPCS 4.0. Not something that should be addressed in this PR.

jrfnl commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+			$error_code = 'InternalUseOnly';
+			$data       = array( $autoload_info['clean'] );
+		} else {
+			$valid_values        = array_map(
+				function ( $value ) {
+					return '`' . $value . '`';
+				},
+				$valid_values
+			);
+			$valid_values_string = implode( ', ', $valid_values );
```
For your consideration:

```suggestion
			$valid_values_string = '`' . implode( '`, `', $valid_values ) . '`';
```

jrfnl commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+				$data
+			);
+
+			if ( $fix ) {
```
```suggestion
			if ( true === $fix ) {
```

jrfnl reviewed (changes requested):
Hi @rodrigoprimo thanks for making those initial updates.

I've now done a proper code review (with the exception of the tests). Hope this helps!

rodrigoprimo committed:
Apply suggestions from code review

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

rodrigoprimo committed:
Remove unused parameter

rodrigoprimo committed:
Rename test case file to be able to add more test case files with syntax errors

rodrigoprimo committed:
Handle case when $array_token is false + add a test for it

rodrigoprimo committed:
Improve sniff XML doc following suggestions from code review

rodrigoprimo committed:
Add link to external blog post about when to use the autoload flag

rodrigoprimo committed:
Add backticks to all function calls in the docblocks

rodrigoprimo committed:
Rename property `valid_values_other_functions` to `valid_values_wp_set_option_functions`

rodrigoprimo committed:
Explain in the docblock the array properties where all values are `true`

rodrigoprimo committed:
Ensure all `@param array` tags explain the structure of the array

rodrigoprimo committed:
Only record `param missing` metric when the autoload parameter is optional

rodrigoprimo commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,102 @@
+    </standard>
+    <code_comparison>
+        <code title="Valid: Using a boolean value or `null`">
+        <![CDATA[
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+ * wp_set_option_autoload(), wp_set_option_autoload_values() are missing the `$autoload` param or
+ * contain an invalid, internal or deprecated value.
+ *
+ * @link https://github.com/WordPress/WordPress-Coding-Standards/issues/2473
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+	const METRIC_NAME = 'Value of the `$autoload` parameter in the option functions';
+
+	/**
+	 * Valid values for the `$autoload` parameter in the add_option() and update_option() functions.
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+	 *
+	 * @var array<string, string>
+	 */
+	private $valid_values_other_functions = array(
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+	 *
+	 * @since 3.2.0
+	 *
+	 * @var array<string, true>
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+			true
+		);
+
+		if ( false === isset( Collections::arrayOpenTokensBC()[ $this->tokens[ $array_token ]['code'] ] ) ) {
```
Done, I also added a test for.

rodrigoprimo commented:
Thanks for taking another look at this PR, @jrfnl. I started replying to your feedback and implementing the changes, but I was unable to finish everything today. I won't be able to look into this next week. The following week, I will address the comments that I was unable to address today.

rodrigoprimo commented:
Rebased without changes to fix an unrelated failing test.

rodrigoprimo committed:
Use array union instead of `array_merge()` to create the list of known discouraged values

rodrigoprimo committed:
Use one error code for fixable internal values and another for non-fixable internal values

rodrigoprimo committed:
Update if condition to explicitly check what is expected and add a test to reach that condition

rodrigoprimo committed:
Add another test without whitespaces surrounding the parameter to test the condition on OptionAutoloadSniff.php:247 correctly

rodrigoprimo committed:
Add a comment explaining why `null` needs special treatment on line 361

rodrigoprimo committed:
Explicitly check if $param_second_token is an integer

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+			$data       = array( $autoload_info['clean'], $this->fixable_values[ $autoload_value ] );
+		} elseif ( isset( $this->internal_values_fixable[ $autoload_value ] ) ) {
+			$message    = 'The use of `%s` as the value of the `$autoload` parameter is discouraged. Use `%s` instead.';
+			$error_code = 'InternalUseOnly';
```
Done

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+
+		$normalized_value = strtolower( $autoload_info['clean'] );
+
+		if ( \T_NS_SEPARATOR === $this->tokens[ $param_first_token ]['code'] && $param_second_token ) {
```
I updated the if condition to explicitly test that `$param_second_token` is an integer and added a test with just `\` as the content of the parameter.

rodrigoprimo commented on WordPress/Tests/WP/OptionAutoloadUnitTest.inc:
```diff
@@ -0,0 +1,229 @@
+ * These are ignored because the option names and the autoload values can't be reliably determined
+ * or the `$options` parameter is empty.
+ */
+$ignored = wp_set_option_autoload_values( $options );
```
I went ahead and added this test. To be honest, it is not clear to me why it is necessary, and this test is testing that was not already being tested.

rodrigoprimo commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,439 @@
+			return;
+		}
+
+		if ( $param_second_token
```
I made the test more explicit by checking if `$param_second_token` is an integer.

rodrigoprimo commented:
> What about moving it into the condition ?

Regarding [the comment quoted above](https://github.com/WordPress/WordPress-Coding-Standards/pull/2520#discussion_r2185578684), I'm not sure why I cannot reply directly to it, but it was addressed, and now the `param missing` metric will only be recorded for `add_option()` and `update_option()`.

rodrigoprimo committed:
Move the third block of the documentation up as suggestion during code review

rodrigoprimo commented:
This will need to be updated to work on PHPCS 4.0 before it is reviewed again.

rodrigoprimo commented:
Force pushed to rebase the branch on the latest `develop` without other changes.

rodrigoprimo committed:
Add tests for namespaced names

rodrigoprimo committed:
Bump since tag as 3.2.0 and 3.3.0 have been released

rodrigoprimo commented:
Just documenting here that I pushed two new commits to this PR:

1. Bump the version number in the `@since` tags
2. Add tests for namespaced names in preparation for PHPCS 4.0

GaryJones commented on WordPress-Extra/ruleset.xml:
```diff
@@ -193,6 +193,13 @@
 		 https://github.com/WordPress/WordPress-Coding-Standards/issues/2459 -->
 	<rule ref="WordPress.WP.GetMetaSingle"/>
 
+	<!-- Flags calls to add_option(), update_option(), wp_set_options_autoload(),
+		 wp_set_option_autoload(), wp_set_option_autoload_values() functions when the
+		 `$autoload` parameter is missing or contain an invalid, internal-only or
+		 deprecated value.
```
```suggestion
	<!-- Flags calls to add_option(), update_option(), wp_set_options_autoload(),
		 wp_set_option_autoload(), wp_set_option_autoload_values() functions when the
		 `$autoload` parameter is missing or contains an invalid, internal-only, or
		 deprecated value.
```

GaryJones commented on WordPress/Docs/WP/OptionAutoloadStandard.xml:
```diff
@@ -0,0 +1,100 @@
+    When using `add_option()`, `update_option()`, `wp_set_options_autoload()`,
+    `wp_set_option_autoload()`, or `wp_set_option_autoload_values()`, it is recommended to
+    explicitly set the autoload parameter to ensure predictable behavior.
+    The autoload flag determines whether the option is automatically loaded on every page load,
```
```suggestion
    The autoload value determines whether the option is automatically loaded on every page load,
```

GaryJones commented on WordPress/Sniffs/WP/OptionAutoloadSniff.php:
```diff
@@ -0,0 +1,442 @@
+	);
+
+	/**
+	 * Internal-use only values for `$autoload` that cannot be fixed automatically by the sniff.
```
```suggestion
	 * Internal-use only values for `$autoload` that will not be fixed automatically by the sniff.
```

GaryJones reviewed (changes requested):
Nothing jumps out at me besides a couple of string fixes.

rodrigoprimo committed:
Apply suggestions from code review

Co-authored-by: Gary Jones <github@garyjones.io>

rodrigoprimo commented:
Thanks for taking the time to review this PR, @GaryJones. I have accepted the three suggestions you made.
